core/tx_list.go

```diff
@@ -0,0 +1,332 @@
+// maintenance.
+func (l *txList) Forward(threshold uint64) types.Transactions {
+	var removed types.Transactions
+
+	if l.strict {
+		// In strict mode, push the lowest nonce forward to the threshold
+		for l.first < threshold {
+			if tx, ok := l.items[l.first]; ok {
+				removed = append(removed, tx)
+			}
+			delete(l.items, l.first)
+			l.first++
+		}
+		if l.first > l.last {
+			l.last = l.first
+		}
+	} else {
+		// In gapped mode, pop off heap items until the threshold is reached
+		for l.index.Len() > 0 && (*l.index)[0] < threshold {
+			nonce := heap.Pop(l.index).(uint64)
+			removed = append(removed, l.items[nonce])
+			delete(l.items, nonce)
+		}
+	}
+	l.cache = nil
+
+	return removed
+}
+
+// Filter removes all transactions from the list with a cost higher than the
+// provided threshold. Every removed transaction is returned for any post-removal
+// maintenance. Strict-mode invalidated transactions are also returned.
+//
+// This method uses the cached costcap to quickly decide if there's even a point
+// in calculating all the costs or if the balance covers all. If the threshold is
+// loewr than the costcap, the costcap will be reset to a new high after removing
+// expensive the too transactions.
+func (l *txList) Filter(threshold *big.Int) (types.Transactions, types.Transactions) {
+	// If all transactions are blow the threshold, short circuit
+	if l.costcap.Cmp(threshold) <= 0 {
+		return nil, nil
+	}
+	l.costcap = new(big.Int).Set(threshold) // Lower the cap to the threshold
+
+	// Gather all the transactions needing deletion
+	var removed types.Transactions
+	for _, tx := range l.items {
+		if cost := tx.Cost(); cost.Cmp(threshold) > 0 {
+			removed = append(removed, tx)
+			delete(l.items, tx.Nonce())
+		}
+	}
+	// Readjust the nonce boundaries/indexes and gather invalidate tranactions
+	var invalids types.Transactions
+	if l.strict {
+		// In strict mode iterate find the first gap and invalidate everything after it
+		for i := l.first; i <= l.last; i++ {
+			if _, ok := l.items[i]; !ok {
+				// Gap found, invalidate all subsequent transactions
+				for j := i + 1; j <= l.last; j++ {
+					if tx, ok := l.items[j]; ok {
+						invalids = append(invalids, tx)
+						delete(l.items, j)
+					}
+				}
+				// Reduce the highest transaction nonce and return
+				l.last = i - 1
+				break
+			}
+		}
+	} else {
+		// In gapped mode no transactions are invalid, but the heap is ruined
+		l.index = &nonceHeap{}
+		for nonce, _ := range l.items {
+			*l.index = append(*l.index, nonce)
+		}
+		heap.Init(l.index)
+	}
+	l.cache = nil
+
+	return removed, invalids
+}
+
+// Cap places a hard limit on the number of items, returning all transactions
+// exceeding tht limit.
+func (l *txList) Cap(threshold int) types.Transactions {
+	// Short circuit if the number of items is under the limit
+	if len(l.items) < threshold {
+		return nil
+	}
+	// Otherwise gather and drop the highest nonce'd transactions
```

Comment:
The highest nonce transactions are farthest from executing. Dropping the lowest nonced ones would create a gap that causes all later transactions to become useless.